nemo/collections/nlp/models/language_modeling/megatron_bert_model.py

```diff
@@ -765,6 +767,36 @@ def configure_optimizers(self):
                     param._disable_greedy_grad_copy = not self.megatron_amp_o2
                     param._disable_overlap_grad_sync = True
 
+            # Initialize parameter buckets for overlapped grad and param syncs
+            buckets = []
+            if self.cfg.get('virtual_pipeline_model_parallel_size', None) is not None:
+                # Initialize a bucket for each virtual pipeline stage
+                for module in self.model:
+                    if isinstance(module, Float16Module):
+                        module = module.module
+                    stage_bucket = []
+                    for layer in module.language_model.encoder.layers:
+                        stage_bucket.extend(
+                            p for p in layer.parameters() if not getattr(p, '_disable_overlap_grad_sync', False)
+                        )
+                    buckets.append(stage_bucket)
+            else:
+                # Initialize a bucket for each Transformer layer
+                modules = self.model if isinstance(self.model, list) else [self.model]
+                for module in modules:
+                    if isinstance(module, Float16Module):
+                        module = module.module
+                    for layer in module.language_model.encoder.layers:
+                        buckets.append(
+                            [p for p in layer.parameters() if not getattr(p, '_disable_overlap_grad_sync', False)]
+                        )
+            buckets.reverse()
+            used_params = set()
+            for bucket in buckets:
+                used_params.update(bucket)
+            buckets.append([p for p in self.parameters() if p not in used_params])
+            self.distributed_adam_buckets = buckets
```

Comment:
This code looks the same for bert and gpt. Can we refactor it so that the code is not duplicated? Maybe we can add a private method to the base class?

Reply:
I have a draft implementation, but I have some misgivings. The issue is that a lot of logic, e.g. for `Float16Module` and interleaved pipelining, is handled in the derived classes, so handling it in the base class breaks encapsulation. Also, T5 doesn't follow the same patterns as GPT and BERT, e.g. it stores the model in `self.enc_dec_model` instead of `self.model`, so handling that behavior will also add messiness. That said, I figure we'll need to rationalize the code structure and move logic to the base class eventually, so perhaps it's worthwhile to get started now.

Reply:
See https://github.com/timmoon10/NeMo/blob/dist-adam-variable-buckets2/nemo/collections/nlp/models/language_modeling/megatron_base_model.py#L399-L456 for a draft implementation.